src/core/RandomVariables.jl

```diff
@@ -118,32 +130,224 @@ mutable struct UntypedVarInfo <: AbstractVarInfo
 end
 VarInfo() = UntypedVarInfo()
 
+###########################
+# Single variable VarInfo #
+###########################
+
+struct SingleVarInfo{sym, T, TDist <: Distribution, TDists <: AbstractVector{TDist}, TVN <: AbstractVector{VarName{sym}}, TVal <: AbstractVector{T}, TRanges <: AbstractVector{UnitRange{Int}}, TId <: AbstractVector{Set{Selector}}, Torders <: AbstractVector{Int}}
+    idcs        ::    Dict{VarName{sym}, Int}
+    vns         ::    TVN
+    ranges      ::    TRanges
+    vals        ::    TVal
+    dists       ::    TDists
+    gids        ::    TId
+    orders      ::    Torders   # observe statements number associated with random variables
+    flags       ::    Dict{String, BitVector}
+end
+SingleVarInfo{sym, T}() where {sym, T} = SingleVarInfo{sym, T, Distribution}()
+function SingleVarInfo{sym, T, TDist}() where {sym, T, TDist}
+    vals  = Vector{T}()
+    flags = Dict{String, BitVector}()
+    flags["del"] = BitVector()
+    flags["trans"] = BitVector()
+
+    SingleVarInfo(
+        Dict{VarName{sym}, Int}(),
+        Vector{VarName{sym}}(),
+        Vector{UnitRange{Int}}(),
+        vals,
+        Vector{TDist}(),
+        Vector{Set{Selector}}(),
+        Vector{Int}(),
+        flags
+    )
+end
+
+function Base.empty!(vi::SingleVarInfo)
+    empty!(vi.idcs)
+    empty!(vi.vns)
+    empty!(vi.ranges)
+    empty!(vi.vals)
+    empty!(vi.dists)
+    empty!(vi.gids)
+    empty!(vi.orders)
+    for k in keys(vi.flags)
+        empty!(vi.flags[k])
+    end
+
+    return vi
+end
+
+#######################
+# Fully typed VarInfo #
+#######################
+
+struct TypedVarInfo{Tvis, Tlogp} <: AbstractVarInfo
+    vis::Tvis
+    logp::Base.RefValue{Tlogp}
+    num_produce::Base.RefValue{Int}
+end
+function TypedVarInfo(vi::UntypedVarInfo)
+    vis = SingleVarInfo[]
+    syms_tuple = Tuple(syms(vi))
+    for s in syms_tuple
+        _inds = findall(vn -> vn.sym == s, vi.vns)
+        sym_inds = collect(1:length(_inds))
+        sym_vns = getindex.((vi.vns,), _inds)
+        sym_idcs = Dict(a => i for (i, a) in enumerate(sym_vns))
+        sym_dists = getindex.((vi.dists,), _inds)
+        sym_gids = getindex.((vi.gids,), _inds)
+        sym_orders = getindex.((vi.orders,), _inds)
+        sym_flags = Dict(a => vi.flags[a][_inds] for a in keys(vi.flags))
+
+        _ranges = getindex.((vi.ranges,), _inds)
+        # `copy` is a workaround to reduce the eltype from Real to Int or Float64
```

Comment:
Do you mean `copy(::Vector{Real})` gives `Vector{Float64}` if each element is `Float64`?

Reply:
`copy.(::Vector{Real})` gives `Vector{Float64}` if each element is `Float64`. note the dot.

Reply:
I see. Great thanks!